### PR TITLE
Automatica2014 SMErobotics reveting demo version

### PR DESCRIPTION
**Pull request by lorenz-halt**
FIX: Somewhere in reading digitals the value was inverted. I change that a high output is read as a high input if there are shortcut.
FIX: If an output is requested to change its state but it is already its state, this version will not return an error.
ADD: An additional srvs that allows checking for exactly one state. Can be used instead of parsing the topic.

This version worked fine on Automatica2014. It is originally based on bnm's fork

### CODE REVIEW

File: cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h [hunk 1]
```diff
@@ -86,9 +86,9 @@ class PhidgetIKROS: public PhidgetIK
 	ros::ServiceServer _srvDigitalOut;
 	ros::ServiceServer _srvDataRate;
 	ros::ServiceServer _srvTriggerValue;
+    ros::ServiceServer _srvDigitalIn;
```
Comment (floweisshardt) [inline cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h]: tabs vs spaces

File: cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h [hunk 2]
```diff
@@ -127,5 +127,9 @@ class PhidgetIKROS: public PhidgetIK
 										cob_phidgets::SetDataRate::Response &res) -> bool;
 	auto setTriggerValueCallback(cob_phidgets::SetTriggerValue::Request &req,
 										cob_phidgets::SetTriggerValue::Response &res) -> bool;
+
+    auto getDigitalCallback(cob_phidgets::SetDigitalSensor::Request &req,
```
Comment (floweisshardt) [inline cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h]: tabs vs spaces

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 3]
```diff
@@ -76,6 +76,8 @@ PhidgetIKROS::PhidgetIKROS(ros::NodeHandle nh, int serial_num, std::string board
 	_srvDataRate = nodeHandle.advertiseService("set_data_rate", &PhidgetIKROS::setDataRateCallback, this);
 	_srvTriggerValue = nodeHandle.advertiseService("set_trigger_value", &PhidgetIKROS::setTriggerValueCallback, this);
 
+    _srvDigitalIn = nodeHandle.advertiseService("get_digital", &PhidgetIKROS::getDigitalCallback, this);
```
Comment (floweisshardt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: tabs vs spaces

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 4]
```diff
@@ -222,7 +224,7 @@ auto PhidgetIKROS::update() -> void
 		if(_indexNameMapItr != _indexNameMapDigitalIn.end())
 			name = (*_indexNameMapItr).second;
 		names.push_back(name);
-		states.push_back(this->getInputState(i));
+        states.push_back(!this->getInputState(i)); // Inverted by the board for unknown reasons
```
Comment (floweisshardt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: tabs vs spaces

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 5]
```diff
@@ -328,7 +330,7 @@ auto PhidgetIKROS::sensorChangeHandler(int index, int sensorValue) -> int
 auto PhidgetIKROS::setDigitalOutCallback(cob_phidgets::SetDigitalSensor::Request &req,
 										cob_phidgets::SetDigitalSensor::Response &res) -> bool
 {
-	bool ret = false;
+    bool ret = false;
```
Comment (floweisshardt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: tabs vs spaces

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 6]
```diff
@@ -340,9 +342,14 @@ auto PhidgetIKROS::setDigitalOutCallback(cob_phidgets::SetDigitalSensor::Request
 	if(_indexNameMapRevItr != _indexNameMapDigitalOutRev.end())
 	{
 		ROS_INFO("Setting digital output %i to state %i", _indexNameMapDigitalOutRev[req.uri], req.state);
-		this->setOutputState(_indexNameMapDigitalOutRev[req.uri], req.state);
+        int lastError = this->setOutputState(_indexNameMapDigitalOutRev[req.uri], req.state);
```
Comment (floweisshardt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: tabs vs spaces

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 7]
```diff
@@ -356,11 +363,12 @@ auto PhidgetIKROS::setDigitalOutCallback(cob_phidgets::SetDigitalSensor::Request
 			ros::Duration(0.025).sleep();
 		}
 		_mutex.lock();
-		res.uri = _indexNameMapDigitalOut[_outputChanged.index];
-		res.state = _outputChanged.state;
+        res.uri = req.uri;
```
Comment (floweisshardt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: tabs vs spaces

**Comment by floweisshardt** [comment]
@ipa-bnm: please review functionality

**Comment by floweisshardt** [comment]
@ipa-bnm: please review. If there is no review until end of January, I'll close this as "won't merge"

**Comment by lorenz-halt** [comment]
The cob_phidget board will be used in EuRoC, and is therefore reviewed again by me.

Please check the functionability against COB and possibly indigo. That I don't need to recommend my fork but the ipa320 base.

File: cob_phidgets/ros/src/phidgetik_ros.cpp (hunk 4 above)
Comment (lorenz-halt) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: If I directly connect input and output, the states are only equal if inversion takes place. However, if nothing is connected the input state is always high (pull-up)

**Comment by fmessmer** [comment]
@ipa-lth please fix the indentation as remarked by @ipa-fmw, i.e. tabs. vs. spaces
You can also rebase this PR to indigo_dev for Indigo support if needed!

Afterwards @ipa-bnm should review functionality in order to give final ok!

File: cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h [hunk 8]
```diff
@@ -86,9 +86,9 @@ class PhidgetIKROS: public PhidgetIK
 	ros::ServiceServer _srvDigitalOut;
 	ros::ServiceServer _srvDataRate;
 	ros::ServiceServer _srvTriggerValue;
+	ros::ServiceServer _srvDigitalIn;
 
 	int _serial_num;
-	std::string _board_name;
```
Comment (benmaidel) [inline cob_phidgets/ros/include/cob_phidgets/phidgetik_ros.h:91]: Please do not remove "_board_name". The boards serial num should only be used as uri path identifierer if no yaml configuration is present. See phidget_sensors_node.cpp line 138 to 145.

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 9]
```diff
@@ -60,7 +60,7 @@
 #include <cob_phidgets/phidgetik_ros.h>
 
 PhidgetIKROS::PhidgetIKROS(ros::NodeHandle nh, int serial_num, std::string board_name, XmlRpc::XmlRpcValue* sensor_params, SensingMode mode)
-	:PhidgetIK(mode), _nh(nh), _serial_num(serial_num), _board_name(board_name)
+	:PhidgetIK(mode), _nh(nh), _serial_num(serial_num)
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 10]
```diff
@@ -147,7 +149,7 @@ auto PhidgetIKROS::readParams(XmlRpc::XmlRpcValue* sensor_params) -> void
 		if(_indexNameMapItr == _indexNameMapDigitalIn.end())
 		{
 			std::stringstream ss;
-			ss << _board_name << "/" << "in/" << i;
+			ss << getDeviceSerialNumber() << "/" << "in/" << i;
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 11]
```diff
@@ -158,7 +160,7 @@ auto PhidgetIKROS::readParams(XmlRpc::XmlRpcValue* sensor_params) -> void
 		if(_indexNameMapItr == _indexNameMapDigitalOut.end())
 		{
 			std::stringstream ss;
-			ss << _board_name << "/" << "out/" << i;
+			ss << getDeviceSerialNumber() << "/" << "out/" << i;
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 12]
```diff
@@ -169,7 +171,7 @@ auto PhidgetIKROS::readParams(XmlRpc::XmlRpcValue* sensor_params) -> void
 		if(_indexNameMapItr == _indexNameMapAnalog.end())
 		{
 			std::stringstream ss;
-			ss << _board_name << "/" << i;
+			ss << getDeviceSerialNumber() << "/" << i;
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 13]
```diff
@@ -182,7 +184,7 @@ auto PhidgetIKROS::readParams(XmlRpc::XmlRpcValue* sensor_params) -> void
 		if(_indexNameMapItr != _indexNameMapDigitalIn.end())
 		{
 			std::stringstream ss;
-			ss << _board_name << "/" << "in/" << i;
+			ss << getDeviceSerialNumber() << "/" << "in/" << i;
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 14]
```diff
@@ -304,7 +306,7 @@ auto PhidgetIKROS::outputChangeHandler(int index, int outputState) -> int
 }
 auto PhidgetIKROS::sensorChangeHandler(int index, int sensorValue) -> int
 {
-	ROS_DEBUG("Board %s: Analog Input %d changed to Value: %d", _board_name.c_str(), index, sensorValue);
+	ROS_DEBUG("Board %d: Analog Input %d changed to Value: %d", _serial_num, index, sensorValue);
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 15]
```diff
@@ -295,7 +297,7 @@ auto PhidgetIKROS::inputChangeHandler(int index, int inputState) -> int
 
 auto PhidgetIKROS::outputChangeHandler(int index, int outputState) -> int
 {
-	ROS_DEBUG("Board %s: Digital Output %d changed to State: %d", _board_name.c_str(), index, outputState);
+	ROS_DEBUG("Board %d: Digital Output %d changed to State: %d", _serial_num, index, outputState);
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 16]
```diff
@@ -273,7 +275,7 @@ auto PhidgetIKROS::update() -> void
 
 auto PhidgetIKROS::inputChangeHandler(int index, int inputState) -> int
 {
-	ROS_DEBUG("Board %s: Digital Input %d changed to State: %d", _board_name.c_str(), index, inputState);
+	ROS_DEBUG("Board %d: Digital Input %d changed to State: %d", _serial_num, index, inputState);
```
Comment (benmaidel) [inline cob_phidgets/ros/src/phidgetik_ros.cpp]: Not OK. See phidgetik_ros.h comment!

**Comment by lorenz-halt** [comment]
Fixed all above comments

File: cob_phidgets/CMakeLists.txt [hunk 17]
```diff
@@ -48,7 +48,12 @@ catkin_package(
 ###########
 ## Build ##
 ###########
-include_directories(${PROJECT_SOURCE_DIR}/ros/include ${PROJECT_SOURCE_DIR}/common/include ${Boost_INCLUDE_DIRS} ${catkin_INCLUDE_DIRS})
+
+# Set to debug build
+#SET(CMAKE_BUILD_TYPE debug)
+
+include_directories(${Boost_INCLUDE_DIRS})
+include_directories(${PROJECT_SOURCE_DIR}/ros/include ${PROJECT_SOURCE_DIR}/common/include)
```
Comment (fmessmer) [inline cob_phidgets/CMakeLists.txt:56]: Why is `${catkin_INCLUDE_DIRS}` removed? Do we need it or not?

File: cob_phidgets/CMakeLists.txt [hunk 18]
```diff
@@ -74,7 +79,13 @@ target_link_libraries(phidget_sensors ${catkin_LIBRARIES})
...
 
 ## Mark executables and/or libraries for installation
-install(TARGETS range_sensors phidget_sensors
+install(TARGETS range_sensors
+  ARCHIVE DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+  LIBRARY DESTINATION ${CATKIN_PACKAGE_LIB_DESTINATION}
+  RUNTIME DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
+
+install(TARGETS phidget_sensors
```
Comment (fmessmer) [inline cob_phidgets/CMakeLists.txt:88]: Why split it up? You can install several target within one `install` command

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 19]
```diff
@@ -222,7 +224,7 @@ auto PhidgetIKROS::update() -> void
 		if(_indexNameMapItr != _indexNameMapDigitalIn.end())
 			name = (*_indexNameMapItr).second;
 		names.push_back(name);
-		states.push_back(this->getInputState(i));
+		states.push_back(!this->getInputState(i)); // Inverted by the board for unknown reasons
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgetik_ros.cpp:227]: So has this been wrong all the time? Or is this only happening on some boards?
@ipa-bnm where on raw's or cob's do we use this/can we verify this?

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 20]
```diff
@@ -273,7 +275,7 @@ auto PhidgetIKROS::update() -> void
 
 auto PhidgetIKROS::inputChangeHandler(int index, int inputState) -> int
 {
-	ROS_DEBUG("Board %s: Digital Input %d changed to State: %d", _board_name.c_str(), index, inputState);
+	ROS_DEBUG("Board %d: Digital Input %d changed to State: %d", _board_name.c_str(), index, inputState);
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgetik_ros.cpp:278]: should be `%s` again as format key...

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 21]
```diff
@@ -295,7 +297,7 @@ auto PhidgetIKROS::inputChangeHandler(int index, int inputState) -> int
 
 auto PhidgetIKROS::outputChangeHandler(int index, int outputState) -> int
 {
-	ROS_DEBUG("Board %s: Digital Output %d changed to State: %d", _board_name.c_str(), index, outputState);
+	ROS_DEBUG("Board %d: Digital Output %d changed to State: %d", _board_name.c_str(), index, outputState);
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgetik_ros.cpp:300]: should be `%s` again as format key...

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 22]
```diff
@@ -304,7 +306,7 @@ auto PhidgetIKROS::outputChangeHandler(int index, int outputState) -> int
 }
 auto PhidgetIKROS::sensorChangeHandler(int index, int sensorValue) -> int
 {
-	ROS_DEBUG("Board %s: Analog Input %d changed to Value: %d", _board_name.c_str(), index, sensorValue);
+	ROS_DEBUG("Board %d: Analog Input %d changed to Value: %d", _board_name.c_str(), index, sensorValue);
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgetik_ros.cpp:309]: should be `%s` again as format key...

File: cob_phidgets/ros/src/phidgetik_ros.cpp [hunk 23]
```diff
@@ -340,7 +342,12 @@ auto PhidgetIKROS::setDigitalOutCallback(cob_phidgets::SetDigitalSensor::Request
 	if(_indexNameMapRevItr != _indexNameMapDigitalOutRev.end())
 	{
 		ROS_INFO("Setting digital output %i to state %i", _indexNameMapDigitalOutRev[req.uri], req.state);
-		this->setOutputState(_indexNameMapDigitalOutRev[req.uri], req.state);
+		int lastError = this->setOutputState(_indexNameMapDigitalOutRev[req.uri], req.state);
+
+		if (lastError != EPHIDGET_OK)
+		{
+			ROS_ERROR("Phidget Error: %i (look up in phidget header file phidget21.h)", lastError);
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgetik_ros.cpp:349]: Is there something like error_to_text in phidget21.h?

File: cob_phidgets/ros/src/phidgets_range_sensors.cpp [hunk 24]
```diff
@@ -21,7 +21,9 @@ class Sensor
 	Sensor(const std::string &fr_id, const int id, const int filter_size = 10) :
 			id_(id), filter_size_(filter_size), frame_id_(fr_id)
 	{
-		pub_range_ = n_.advertise<sensor_msgs::Range>(frame_id_, 0);
+		char buffer[256];
+		sprintf(buffer, "range_%d", id);
+		pub_range_ = n_.advertise<sensor_msgs::Range>(buffer, 0);
```
Comment (fmessmer) [inline cob_phidgets/ros/src/phidgets_range_sensors.cpp:26]: Does this change the name of the topic for cob's or raw's?
Isn't `frame_id` configurable via parameters?
@ipa-lth wouldn't the parameter be an option for you?

**Comment by fmessmer** [comment]
In the end, will this PR go to hydro or indigo?

**Comment by lorenz-halt** [comment]
Closing request. ipa-bnm is informed about the inversion issue.

**Pull request closed without merging**